Port CPU torch.ormqr to ATen

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54725 Added cuBLAS path for torch.linalg.lstsq
* #57317 Add cuSOLVER path for torch.linalg.lstsq
* #57316 Add CUDA support for torch.ormqr
* **#57315 Port CPU torch.ormqr to ATen**

This PR ports `torch.ormqr` from TH to ATen.
CUDA path will be implemented in a follow-up PR.
With ATen port, support for complex and batched inputs is added.
The tests are rewritten and OpInfo entry is added.

We can implement the least squares solver with geqrf + ormqr +
triangular_solve. So it's useful to have this function renewed at least for the
internal code.

Resolves https://github.com/pytorch/pytorch/issues/24748

Differential Revision: [D28242070](https://our.internmc.facebook.com/intern/diff/D28242070)

IvanYashchuk committed:
Port CPU torch.ormqr to ATen

This PR ports `torch.ormqr` from TH to ATen.
CUDA path will be implemented in a follow-up PR.
With ATen port, support for complex and batched inputs is added.
The tests are rewritten and OpInfo entry is added.

We can implement the least squares solver with geqrf + ormqr +
triangular_solve. So it's useful to have this function renewed at least for the
internal code.

Resolves https://github.com/pytorch/pytorch/issues/24748

[ghstack-poisoned]

mruberry reviewed (approved):
Awesome! @Lezcano would you like to sanity check this, too?

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1872,6 +1894,149 @@ Tensor orgqr(const Tensor& input, const Tensor& tau) {
+  TORCH_CHECK(other.dim() >= 2, "torch.ormqr: other must have at least 2 dimensions.");
+
+  int64_t left_size_condition = left ? -2 : -1;
+  TORCH_CHECK(
```
missing `input.size(-2) >= input.size(-1)`?

IvanYashchuk replied:
No, as you see from the tests passing this function works both for `m >= n` and `m < n` matrices.
For the case with `m < n` only the first _m_ columns that represent _m_ Householder vectors are used by LAPACK.

```python
In [1]: import torch
In [2]: a = torch.randn(3, 5)
In [3]: h, tau = torch.geqrf(a)
In [4]: h.shape
Out[4]: torch.Size([3, 5])
In [5]: c = torch.randn(3, 7)
In [6]: res = torch.ormqr(h, tau, c)
In [7]: q, _ = torch.linalg.qr(a)
In [8]: torch.allclose(q @ c, res)
Out[8]: True
```

Lezcano replied:
Then we should add that the case `m < n` to the docs.

IvanYashchuk replied:
Why do you think the case `m < n` is special?

Lezcano replied:
Because in `householder_product` is not even considered:
https://pytorch.org/docs/master/generated/torch.linalg.householder_product.html

IvanYashchuk replied:
Well, `householder_product` is related but it's a different function. It has the same constraints on the input as the original orgqr implementation had:
https://github.com/pytorch/pytorch/blob/27f7d1c2865355c694fb964609df45974748615b/aten/src/TH/generic/THTensorLapack.cpp#L341
and LAPACK's orgqr has this `m>=n` requirement, but n there is the number of columns of Q to be computed, not `n=input.shape[-1]`. It has a side effect that the output of `torch.geqrf` can't be directly used with `torch.orgqr`, but it can be used with `ormqr` with the current implementation.

```python
In [1]: import torch
In [2]: a = torch.randn(3, 5)
In [3]: h, tau = torch.geqrf(a)
In [4]: c = torch.eye(3)
In [5]: torch.ormqr(h, tau, c) # narrow of `h` is not required
Out[5]: 
tensor([[-0.1468,  0.8266, -0.5433],
        [-0.0029, -0.5496, -0.8354],
        [-0.9892, -0.1211,  0.0831]])
In [6]: torch.linalg.qr(a)[0]
Out[6]: 
tensor([[-0.1468,  0.8266, -0.5433],
        [-0.0029, -0.5496, -0.8354],
        [-0.9892, -0.1211,  0.0831]])
In [7]: torch.linalg.householder_product(h, tau)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-7-70a58677659d> in <module>
----> 1 torch.linalg.householder_product(h, tau)

RuntimeError: torch.linalg.householder_product: input.shape[-2] must be greater than or equal to input.shape[-1]
In [8]: torch.linalg.householder_product(h.narrow(-1, 0, 3), tau) # narrow is required here
Out[8]: 
tensor([[-0.1468,  0.8266, -0.5433],
        [-0.0029, -0.5496, -0.8354],
        [-0.9892, -0.1211,  0.0831]])
```

Lezcano replied:
I do not have any strong opinions on how to handle this, as I really think that these functions are too low level to form part of the python API.

That being said, it's quite annoying that we have an `h, tau` decomposition in several functions, and that each of them has slightly different requirements.

Lezcano replied:
Even more, the `seealso` section of `householder_reflection` should be updated to reflect this dissonance.

IvanYashchuk replied:
In every place where we have a multiplication of Q with some other matrix, using this function should be more efficient. In PyTorch Python code, there is for example one place it could be used, in the implementation of `lobpcg` the Q matrix is explicitly formed and then used in multiplication:
https://github.com/pytorch/pytorch/blob/c371542efc31b1abfe6f388042aa3ab0cef935f2/torch/_lobpcg.py#L878
https://github.com/pytorch/pytorch/blob/c371542efc31b1abfe6f388042aa3ab0cef935f2/torch/_linalg_utils.py#L80-L88

To make using this function more accessible, the documentation should be improved, the name should be changed and autograd support added. If we do that, we can revisit the constraints on the input and make it consistent.

Lezcano replied:
That's spot on. Let's do that in a follow-up after the branch-cut.

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1872,6 +1894,149 @@ Tensor orgqr(const Tensor& input, const Tensor& tau) {
+  copy_needed |= (result.numel() != 0 && !result_equal_expected_shape); // or result does not have the expected shape
+  // we have to allocate a temporary tensor
+  if (copy_needed) {
+    Tensor result_tmp = at::empty({0}, input.options());
```
Why you create a new one, then resize the result and then copy, rather than resizing and passing the resized tensor?
With this other logic, this if/else could be rewritten as:
```c++
if (copy_needed) {
    at::native::resize_output(result, result_tmp.sizes());
}
ormqr_out_helper(input, tau, other, result, left, transpose);
```

IvanYashchuk replied:
Passed-in "out" tensor can be resized with `at::native::resize_output`, but resizing non-empty tensors is deprecated. This resizing will be turned into errors later.
If the shape is correct but strides are not, I can't restride it, strides should not be touched, therefore we need a copy with a temporary Fortran-contiguous tensor.

Lezcano replied:
I'm sorry, but I still don't understand this. Two lines below you are using `resize_output` on `result`. Does that mean that if `result` is non-empty, that line will throw an error when that resizing behaviour is deprecated?

IvanYashchuk replied:
> Does that mean that if result is non-empty, that line will throw an error when that resizing behaviour is deprecated?

Yes, it will throw an error for non-empty result tensors that do not match the expected shape. The following warning is included in `at::native::resize_output` and will be changed to an error at some point
```cpp
  if (output.numel() != 0) {
    TORCH_WARN(
      "An output with one or more elements was resized since it had ",
      "shape ", output.sizes(), ", which does not match the required ",
      "output shape ", shape, ".",
      "This behavior is deprecated, and in a future PyTorch release outputs ",
      "will not be resized unless they have zero elements. You can explicitly ",
      "reuse an out tensor t by resizing it, inplace, to zero elements with ",
      "t.resize_(0).");
  }
```
from `aten/src/ATen/native/Resize.cpp`.

Lezcano replied:
Ivan explained this to me offline. The code is correct as it is because:
- If it's the wrong size, we handle this case, but it will throw an error in the future
- If it is the right size but wrong strides, we create a new one with the correct strides, then use that one and copy back to the original one.

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1872,6 +1894,149 @@ Tensor orgqr(const Tensor& input, const Tensor& tau) {
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.device() == other.device());
+
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(result.scalar_type() == input.scalar_type());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(result.device() == input.device());
```
How come all these asserts are here and then repeated in `ormqr_out`?

IvanYashchuk replied:
I keep the `ormqr_out_helper` function only for inputs that are Fortran-contiguous and then `ormqr_out` function decides whether to pass further the "out" tensor if it satisfies the requirements or make a temporary tensor + copy.

Those asserts here are just another level of "sanity" checks as this `ormqr_out_helper` function could in principle be used also in other places (like a least squares solver). If we would accept only Fortran-contiguous "out" tensors, then these two functions would be merged.

lezcano reviewed (commented):
Left a few comments, none of them too important.

IvanYashchuk commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2004,6 +2004,20 @@ def sample_inputs_householder_product(op_info, device, dtype, requires_grad, **k
+    batches = [(), (0, ), (2, ), (2, 1)]
+    ns = [5, 2, 0]
+    tf = [True, False]
+    for batch, (m, n), left, transpose in product(batches, product(ns, ns), tf, tf):
```
The iteration includes also non-square cases:
```python
In [1]: import itertools
In [2]: list(itertools.product([5, 2, 0], [5, 2, 0]))
Out[2]: [(5, 5), (5, 2), (5, 0), (2, 5), (2, 2), (2, 0), (0, 5), (0, 2), (0, 0)]
```

Lezcano replied:
ooops, I don't know what I was thinking about.

IvanYashchuk commented on test/test_linalg.py:
```diff
@@ -4880,48 +4880,61 @@ def renorm(matrix, value, dim, max_norm):
+        batches = [(), (0, ), (2, ), (2, 1)]
+        ns = [5, 2, 0]
+        for batch, (m, n), fortran_contiguous in product(batches, product(ns, ns), [True, False]):
+            run_test(batch, m, n, fortran_contiguous)
```
Non-square matrices are there; in the iteration there are cases with `m!=n`.
```python
In [1]: import itertools
In [2]: list(itertools.product([5, 2, 0], [5, 2, 0]))
Out[2]: [(5, 5), (5, 2), (5, 0), (2, 5), (2, 2), (2, 0), (0, 5), (0, 2), (0, 0)]
```

IvanYashchuk commented on torch/_torch_docs.py:
```diff
@@ -6725,22 +6725,37 @@ def merge_dicts(*dicts):
-This directly calls the underlying LAPACK function `?ormqr`.
-See `LAPACK documentation for ormqr`_ for further details.
+Multiplies a matrix `C` (given by :attr:`other`) with a matrix `Q`,
+where `Q` is represented using Householder reflectors `(input, tau)`.
```
If some elements in `tau` are zeros then that means the corresponding Householder vector will be zeroed, so its contribution will not be considered, leaving an identity transformation instead of reflection. If all elements are zero then the implicit Q matrix will be identity. I think this follows from the LAPACK note on the Householder representation and somebody working with this representation should expect such behavior after reading it.

Lezcano replied:
This is too niche of a function, so the details of the documentation can be just a link to LAPACK, as you already did. Most probably the correct question would be "why are we exposing a function as specific as this one", but that's a point for some other time.

IvanYashchuk committed:
Update on "Port CPU torch.ormqr to ATen"


This PR ports `torch.ormqr` from TH to ATen.
CUDA path will be implemented in a follow-up PR.
With ATen port, support for complex and batched inputs is added.
The tests are rewritten and OpInfo entry is added.

We can implement the least squares solver with geqrf + ormqr +
triangular_solve. So it's useful to have this function renewed at least for the
internal code.

Resolves https://github.com/pytorch/pytorch/issues/24748

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port CPU torch.ormqr to ATen"


This PR ports `torch.ormqr` from TH to ATen.
CUDA path will be implemented in a follow-up PR.
With ATen port, support for complex and batched inputs is added.
The tests are rewritten and OpInfo entry is added.

We can implement the least squares solver with geqrf + ormqr +
triangular_solve. So it's useful to have this function renewed at least for the
internal code.

Resolves https://github.com/pytorch/pytorch/issues/24748

[ghstack-poisoned]

IvanYashchuk commented:
@Lezcano, thank you for your feedback! I've updated this PR according to your suggestions and I expanded a bit the documentation on the sizes of inputs added a "Raises:" section, could you please take a look?

Lezcano commented on torch/_torch_docs.py:
```diff
@@ -6699,11 +6699,13 @@ def merge_dicts(*dicts):
 See `Representation of Orthogonal or Unitary Matrices`_ for further details.
 
 If :attr:`left` is `True` then `op(Q)` times `C` is computed, otherwise the result is `C` times `op(Q)`.
+When ``left==True`` the implicit matrix `Q` has the size  :math:`m \times m`,
```
Perhaps a better wording:

> When :attr:`left` is `True`, the implicit matrix `Q` has size :math:`m \times m`. It has size :math:`n \times n` otherwise.

Lezcano commented on torch/_torch_docs.py:
```diff
@@ -6715,15 +6717,23 @@ def merge_dicts(*dicts):
 Args:
-    input (Tensor): tensor of shape `(*, m, k)` where `*` is zero or more batch dimensions.
-    tau (Tensor): tensor of shape `(*, min(m, k))` where `*` is zero or more batch dimensions.
+    input (Tensor): tensor of shape `(*, mn, k)` where `*` is zero or more batch dimensions.
```
What is `mn`? It has not been defined.
It might be interesting to write a full definition of what's going on at the beginning, as in
https://pytorch.org/docs/master/generated/torch.linalg.householder_product.html

It would even be reasonable to put this function in the `seealso` of `torch.linalg.householder_product`, as this function computes the action of the `householder_product` on a matrix (similar to `inv` vs `solve`) and it should often be preferred over `householder_product` + `matmul`.

IvanYashchuk replied:
> What is mn? It has not been defined.

I'll change to
```
    input (Tensor): tensor of shape `(*, mn, k)` where `*` is zero or more batch dimensions
                    and `mn` is `m` or `n` depending on the :attr:`left`.
```

Lezcano commented on torch/_torch_docs.py:
```diff
@@ -6715,15 +6717,23 @@ def merge_dicts(*dicts):
 
+Raises:
+    RuntimeError:
+        if :attr:`input` and :attr:`other` have incompatible shapes for the matrix multiplication.
```
We are not reporting the size errors in this section, as basically every function has some "size compatibility requirements" and it'd mean too much repetition.

IvanYashchuk replied:
Okay, I'll remove this section then and also the note about tau as it's also a size incompatibility issue.

Lezcano commented on torch/_torch_docs.py:
```diff
@@ -6715,15 +6717,23 @@ def merge_dicts(*dicts):
+        if :attr:`input` and :attr:`other` have incompatible shapes for the matrix multiplication.
+
+    RuntimeError:
+        if size of :attr:`tau` vector is larger than number of rows (when ``left=True``)
```
nit. "if the size of :attr:`tau` is larger than the number", or even better, "the length".
Also, without double quotes, it should be `` `:attr:`left`\ `= True` `` (same in the next line).

lezcano reviewed (approved):
I just left a couple comments. Just stylistic points.
In any case, this PR is ready to be merged.

IvanYashchuk committed:
Update on "Port CPU torch.ormqr to ATen"


This PR ports `torch.ormqr` from TH to ATen.
CUDA path will be implemented in a follow-up PR.
With ATen port, support for complex and batched inputs is added.
The tests are rewritten and OpInfo entry is added.

We can implement the least squares solver with geqrf + ormqr +
triangular_solve. So it's useful to have this function renewed at least for the
internal code.

Resolves https://github.com/pytorch/pytorch/issues/24748

[ghstack-poisoned]

IvanYashchuk commented:
@mruberry, I think this stack is ready to be merged. Could you please take another look?

I also moved the cuBLAS path for lstsq PR to this stack after the cuSOLVER PR.

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28242070).

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28242070).
